devito/types/basic.py

```diff
@@ -879,6 +879,14 @@ def __new__(cls, *args, **kwargs):
             # let's just return `function` itself
             return function
 
+        # If dimensions have been replaced, then it is necessary to set `function`
+        # to None. It is also necessary to remove halo and padding kwargs so that
+        # they are rebuilt with the new dimensions
+        if function is not None and function.dimensions != dimensions:
```

Comment:
I'm sorry but no we cannot allow that it doesn't matter if it's for internals, this need to go through a `__new__` not `_rebuild`

Reply:
But `_rebuild()` is a `__new__` with some extra convenience around it?